### ISSUE TITLE
[1.84.0 regression] `#[allow(clippy::disallowed_macros)]` does not work on two-argument `std::panic`

### ISSUE THREAD

**Issue by Deewiant** [rust-lang/rust-clippy#14017]
### Summary

As of 1.84.0, it is not possible to use `#[allow]` or `#[expect]` to silence `clippy::disallowed_macros` when invoking `std::panic` with exactly two arguments.

Or at least, the straightforward thing does not work, and I haven't found a workaround other than "don't use `std::panic` with two arguments".

As far as I can tell, this is specific to `std::panic` and specific to the two-argument case, which seems pretty strange... But I obviously haven't exhaustively tested all macros so maybe that is just a coincidence.

### Lint Name

disallowed_macros

### Reproducer

I tried this code:

```rust
pub fn print_1_arg() {
    #![expect(clippy::disallowed_macros)]
    print!("foo");
}
pub fn print_2_args() {
    #![expect(clippy::disallowed_macros)]
    print!("{}", 123);
}
pub fn print_3_args() {
    #![expect(clippy::disallowed_macros)]
    print!("{} {}", 123, 456);
}

pub fn panic_1_arg() {
    #![expect(clippy::disallowed_macros)]
    panic!("foo");
}
pub fn panic_2_args() {
    #![expect(clippy::disallowed_macros)]
    panic!("{}", 123);
}
pub fn panic_3_args() {
    #![expect(clippy::disallowed_macros)]
    panic!("{} {}", 123, 456);
}
```

With this `clippy.toml`:

```toml
disallowed-macros = ["std::panic", "std::print"]
```

I saw this happen:

```
warning: use of a disallowed macro `std::panic`
  --> asdf.rs:20:5
   |
20 |     panic!("{}", 123);
   |     ^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_macros
   = note: `#[warn(clippy::disallowed_macros)]` on by default

warning: this lint expectation is unfulfilled
  --> asdf.rs:19:15
   |
19 |     #![expect(clippy::disallowed_macros)]
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unfulfilled_lint_expectations)]` on by default
```

I expected to see this happen: no error.

With `cargo +1.83.0 clippy` it works, with `cargo +1.84.0 clippy` it no longer does. (And at least 1.82.0 and 1.81.0 also work; didn't test older than that.)

If the `#![expect]` lines are removed, I see six errors with both versions as expected — though 1.84.0 strangely reports the two-argument `std::panic` one, i.e. the problematic one that can't be silenced, as the very first error, instead of reporting the errors in source code order. Maybe that's somehow related?

### Version

```text
rustc 1.84.0 (9fc6b4312 2025-01-07)
binary: rustc
commit-hash: 9fc6b43126469e3858e2fe86cafb4f0fd5068869
commit-date: 2025-01-07
host: x86_64-unknown-linux-gnu
release: 1.84.0
LLVM version: 19.1.5
```

**Comment by 2asoft** [comment]
Bisecting the the https://github.com/rust-lang/rust repo with the repro above points to https://github.com/rust-lang/rust/pull/132746.  Reviewing its content, this stood out: https://github.com/rust-lang/rust-clippy/pull/13657.

Adding the tests as in the description and bisecting this (rust-clippy) repo between tags `rust-1.83.0` and `rust-1.84.0`, the merge commit for the above clippy PR was the first one that caused the lint warnings to become reordered, as this issue describes.  I pushed a branch with the tests: https://github.com/2asoft/rust-clippy/commits/2asoft/disallowed_macros/.

Curiously:
- If the format string in `panic!("{}", 123);` is modified to be literally anything else (within valid format string limits), the lint warnings are no longer reordered AND the lint is applied correctly (ie the test in my branch passes, and the expectation as described in the repro in this issue is also fulfilled).  For example, `panic!("{} ", 123);`, note the additional space in the string.
- The lint applies to `panic!();` the same as it does to `panic!("{}", 123);`; that is, it is also reordered above others, after the above clippy PR was merged.
- All of this is applicable only to edition 2021 onwards.


The reason for this is that the PR in question changed the order in which attributes are processed.  Previously, `LateLintPass` worked over all AST nodes in order it processes them.  Now, `EarlyLintPass`-discovered attributes are processed first, then the rest of `LateLintPass`.
This matters because of how the [`panic!` macro expands in edition 2021](https://github.com/rust-lang/rust/blob/9fc6b43126469e3858e2fe86cafb4f0fd5068869/library/core/src/panic.rs#L80):
```rs
#[doc(hidden)]
#[unstable(feature = "edition_panic", issue = "none", reason = "use panic!() instead")]
#[allow_internal_unstable(
    panic_internals,
    core_intrinsics,
    const_dispatch,
    const_eval_select,
    const_format_args,
    rustc_attrs
)]
#[rustc_diagnostic_item = "core_panic_2021_macro"]
#[rustc_macro_transparency = "semitransparent"]
#[cfg(not(feature = "panic_immediate_abort"))]
pub macro panic_2021 {
    () => ({
        // Create a function so that the argument for `track_caller`
        // can be moved inside if possible.
        #[cold]
        #[track_caller]
        #[inline(never)]
        const fn panic_cold_explicit() -> ! {
            $crate::panicking::panic_explicit()
        }
        panic_cold_explicit();
    }),
    // Special-case the single-argument case for const_panic.
    ("{}", $arg:expr $(,)?) => ({
        #[cold]
        #[track_caller]
        #[inline(never)]
        #[rustc_const_panic_str] // enforce a &&str argument in const-check and hook this by const-eval
        #[rustc_do_not_const_check] // hooked by const-eval
        const fn panic_cold_display<T: $crate::fmt::Display>(arg: &T) -> ! {
            $crate::panicking::panic_display(arg)
        }
        panic_cold_display(&$arg);
    }),
    ($($t:tt)+) => ({
        // Semicolon to prevent temporaries inside the formatting machinery from
        // being considered alive in the caller after the panic_fmt call.
        $crate::panicking::panic_fmt($crate::const_format_args!($($t)+));
    }),
}
```
The no-arguments and single-argument cases are different, and have attributes applied to them.  This can also be verified by debugging (I just added logging to clippy, for simplicity) clippy to observe the `earlies` in the PR in question contains the attributes pointing to the `panic_2021!` lines for single-argument expansion.
eg on my machine:
```
dbg: in DisallowedMacros::check_crate, crate[earlies=Some(
    [
        tests/ui-toml/disallowed_macros/disallowed_macros.rs:4:1: 4:18 (#0),
        no-location (#1),
        no-location (#1),
        /home/aasoft/.rustup/toolchains/nightly-2024-10-18-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic.rs:94:9: 94:16 (#18),
        /home/aasoft/.rustup/toolchains/nightly-2024-10-18-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic.rs:95:9: 95:24 (#18),
        /home/aasoft/.rustup/toolchains/nightly-2024-10-18-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic.rs:96:9: 96:25 (#18),
        /home/aasoft/.rustup/toolchains/nightly-2024-10-18-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic.rs:97:9: 97:33 (#18),
        /home/aasoft/.rustup/toolchains/nightly-2024-10-18-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic.rs:98:9: 98:36 (#18),
    ],
)]
```

This explains the changed ordering of clippy lint warnings.

---

Digging into it deeper, the PR in question also changed how the attribute warning are emitted.

[Before](https://github.com/rust-lang/rust-clippy/pull/13657/files#diff-d40458190c8bdf76909de7dc9cc6a76ef44092f9a1c4c924b249ab843410ecd0L169), they were emitted within `LateLintPass::check_attribute` for `DisallowedMacros`.
[After](https://github.com/rust-lang/rust-clippy/pull/13657/files#diff-d40458190c8bdf76909de7dc9cc6a76ef44092f9a1c4c924b249ab843410ecd0R127), they are emitted within `LateLintPass::check_crate` for `DisallowedMacros`.

This looks fine on the surface, and there is no difference in text emitted (ignoring the order, at least).  But internally, `rustc_lint` tracks the `HirId` when calling [`LateLintPass::check_attribute`](https://github.com/rust-lang/rust/blob/9fc6b43126469e3858e2fe86cafb4f0fd5068869/compiler/rustc_lint/src/late.rs#L54), [`LateLintPass::check_crate`](https://github.com/rust-lang/rust/blob/9fc6b43126469e3858e2fe86cafb4f0fd5068869/compiler/rustc_lint/src/late.rs#L445), and others.  `DisallowedMacros::check` calls `span_lint_and_then` -> `LateContext::span_lint` -> [`LateContext::opt_span_lint`](https://github.com/rust-lang/rust/blob/9fc6b43126469e3858e2fe86cafb4f0fd5068869/compiler/rustc_lint/src/context.rs#L667), which uses the `HirId` that `rustc_lint` tracks.

I read this to mean that, in effect, by moving the warning emitter into `check_crate`, the warning is applied to the entire mod, not the span on which it originated.

Here's a [change](https://github.com/2asoft/rust/commits/aasoft/i14017/) to log the `HirId` of the emitted warnings, which gave me this:
```
dbg: LateContext::opt_span_lint[lint=Lint {
    name: "clippy::DISALLOWED_MACROS",
    default_level: Warn,
    desc: "use of a disallowed macro",
    edition_lint_opts: None,
    report_in_external_macro: true,
    future_incompatible: None,
    is_externally_loaded: true,
    feature_gate: None,
    crate_level_only: false,
    eval_always: false,
}, hir_id=HirId(DefId(0:0 ~ test_crate[92ea]).0)]
warning: use of a disallowed macro `std::panic`
  --> src/lib.rs:16:5
   |
16 |     panic!();
   |     ^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_macros
   = note: `#[warn(clippy::disallowed_macros)]` on by default

dbg: LateContext::opt_span_lint[lint=Lint {
    name: "clippy::DISALLOWED_MACROS",
    default_level: Warn,
    desc: "use of a disallowed macro",
    edition_lint_opts: None,
    report_in_external_macro: true,
    future_incompatible: None,
    is_externally_loaded: true,
    feature_gate: None,
    crate_level_only: false,
    eval_always: false,
}, hir_id=HirId(DefId(0:0 ~ test_crate[92ea]).0)]
warning: use of a disallowed macro `std::panic`
  --> src/lib.rs:24:5
   |
24 |     panic!("{}", 123);
   |     ^^^^^^^^^^^^^^^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_macros

```

Which I believe confirms my understanding above.

---

@jdonszelmann as the author of the original PR.  What are your thoughts?

**Comment by droundy** [comment]
To be clear, it seems that `#[allow(clippy::disallowed_macros)]` is entirely broken at the moment?

**edit** From testing, it seems that `#![allow(clippy::disallowed_macros)]` works in `src/lib.rs` but otherwise there doesn't seem to be any way to allow the forbidden macros in a specific subset of code.  So I guess the workaround would be to put wrapper code into a separate crate?

**Comment by 2asoft** [comment]
I should have been more explicit in the above.  `#[allow(clippy::disallowed_macros)]` is broken in the specific way described in this issue[1] for scenarios where macro expansion produces code with attributes.  This lint continues to work correctly for other macros.
I have not looked for workarounds. What you wrote looks valid to me based on my exploration above, though I am not sure if it can be applied in all situations.

[1] There is a separate and different issue with how it applies to proc macros: https://github.com/rust-lang/rust-clippy/issues/13521

Of course, the solution is to fix the bug, so the workaround would not be necessary.

**Comment by jalil-salame** [comment]
This seems to also be triggered by the `std::matches!` macro since it has the `#[allow(non_exhaustive_omitted_patterns)]` attribute since Rust 1.90.0:

```rust
//! src/lib.rs

pub enum BoxState {
    Empty,
    Full,
}

impl BoxState {
    #[allow(clippy::disallowed_macros)]
    pub fn is_empty(&self) -> bool {
        matches!(self, BoxState::Empty)
    }

    pub fn is_full(&self) -> bool {
        matches!(self, BoxState::Full)
    }
}
```

```toml
# clippy.toml
disallowed-macros = [{ path = "std::matches" }]
```

```console
❯ cargo +1.89.0 clippy --message-format=short
    Checking bug-report v0.1.0 (/tmp/bug-report)
src/lib.rs:13:9: warning: use of a disallowed macro `std::matches`
warning: `bug-report` (lib) generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.09s
❯ cargo +1.90.0 clippy --message-format=short
    Checking bug-report v0.1.0 (/tmp/bug-report)
src/lib.rs:9:9: warning: use of a disallowed macro `std::matches`
src/lib.rs:13:9: warning: use of a disallowed macro `std::matches`
warning: `bug-report` (lib) generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.18s
❯ cargo +nightly clippy --message-format=short
    Checking bug-report v0.1.0 (/tmp/bug-report)
src/lib.rs:9:9: warning: use of a disallowed macro `std::matches`
src/lib.rs:13:9: warning: use of a disallowed macro `std::matches`
warning: `bug-report` (lib) generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.20s
❯ cargo +nightly --version
cargo 1.96.0-nightly (a357df4c2 2026-04-03)
```

**Comment by 2asoft** [comment]
`std::matches!` and `std::panic!` and any other macros whose expansion introduces an attribute are subject to this.  In my testing, the PR attached to this resolves the issue completely.
